API endpoint to create tenants and dashboards

evamillan opened the pull request:
This PR adds the `/api/create_tenant` endpoint to create a tenant and its associated roles and import the default dashboards found in the `default_dashboards.zip` file into the new tenant.

Sample request:
```
curl -X POST <OSD URL>/api/create_tenant \
--header 'osd-xsrf: true' \
--user '<USERNAME>:<PASSWORD>' \
--header 'Content-Type: application/json' \
-d '{
  "tenant": "new_tenant_name",
  "anonymous": true,
  "force": false
}'
```

The tests for this API are run with `yarn test:jest_server`. They need an OpenSearch server with the security plugin installed running on the background (it can be launched running `yarn opensearch snapshot --security -E discovery.type=single-node` on the OpenSearch Dashboards root directory).

evamillan commented:
I updated the PR to return an error if the tenant or a role already exists and `force` is set to `false`. In that case, the tenant and any role that was created during the execution are deleted.

sduenas commented on server/routes/tenants/create_tenant.ts:
```diff
@@ -0,0 +1,319 @@
+import { IBasePath, IRouter } from '../../../../../src/core/server';
+import { getRolesMapping } from './roles_mapping';
+
+export function registerTenantRoute(router: IRouter, basePath: IBasePath) {
```
Shouldn't it be better to name this `registerCreateTenantRoute`?

sduenas commented on test/api/tenants.test.ts:
```diff
@@ -0,0 +1,175 @@
+  });
+
+  it('Creates anonymous role', async () => {
+    // Delete 'bap_plugins_visibility' role from last test
```
Are tests always run in the same order, one after another? Normally tests should be independent from other tests.

evamillan replied:
I changed it to clean up the database after each test so they are independent from each other.

sduenas commented on server/routes/tenants/create_tenant.ts:
```diff
@@ -0,0 +1,319 @@
+import { IBasePath, IRouter } from '../../../../../src/core/server';
+import { getRolesMapping } from './roles_mapping';
+
+export function registerTenantRoute(router: IRouter, basePath: IBasePath) {
```
Shouldn't it be better to name this registerCreateTenantRoute? I guess each route is for one and only one API endpoint, right?

evamillan replied:
Yes, that is a more accurate name. I've renamed it.

sduenas commented on test/api/tenants.test.ts:
```diff
@@ -0,0 +1,175 @@
+        force: true,
+        tenant: 'new_tenant_name',
+      });
+
```
I miss a test that cleans up everything when the roles already exist.

evamillan replied:
I added a test that checks that no tenant or roles are created if one of the roles already exists.

sduenas reviewed (changes requested):
In overall, I think the PR looks good. There are some small comments we can address. I would like @zhquan to have it a look also to check if all the functionality we need is available here.

zhquan commented:
Using this command I can create all roles for that tenant with all dashboards from https://github.com/chaoss/grimoirelab-sigils/tree/master/panels/json/opensearch_dashboards

```
curl -X POST <OSD URL>/api/create_tenant \
--header 'osd-xsrf: true' \
--user '<USERNAME>:<PASSWORD>' \
--header 'Content-Type: application/json' \
-d '{
  "tenant": "new_tenant_name",
  "anonymous": true,
  "force": false
}'
```
LGTM

evamillan commented:
I updated the PR to add an endpoint to delete the tenants and changed the API paths to make them more semantic.

Sample request to delete a tenant:
```
curl -X DELETE <OSD URL>/api/_bap/tenant/<TENANT NAME> \
--header 'osd-xsrf: true' \
--user '<USERNAME>:<PASSWORD>'
```

Updated request to create a tenant:
```
curl -X POST <OSD URL>/api/_bap/tenant/<TENANT NAME> \
--header 'osd-xsrf: true' \
--user '<USERNAME>:<PASSWORD>' \
--header 'Content-Type: application/json' \
-d '{
  "anonymous": true,
  "force": false
}'
```

evamillan commented:
I added the `deleteIndices` parameter to conditionally delete the tenant's indices:

```
curl -X DELETE <OSD URL>/api/_bap/tenant/<TENANT NAME> \
--header 'osd-xsrf: true' \
--user '<USERNAME>:<PASSWORD>' \
--header 'Content-Type: application/json' \
-d '{
  "deleteIndices": true
}'
```

zhquan commented:
The `deleteIndices` option works perfectly

I think it would be a good option to add the indices to remove because by default the roles allow these indices schemas (the `deleteIndices` option only removes `bap_<tenant>_*`, this should be the by default behavior)

- `grimoirelab_<tenant>_*`
- `bap_<tenant>_*`
- `custom_<tenant>_*`
- `c_<tenant>_*`

For example
- With indices
```
curl -X DELETE <OSD URL>/api/_bap/tenant/<TENANT NAME> \
--header 'osd-xsrf: true' \
--user '<USERNAME>:<PASSWORD>' \
--header 'Content-Type: application/json' \
-d '{
  "deleteIndices": true,
  "indices": "bap_<tenant>_*,c_<tenant>_*,git_enriched"
}'
```
- By default (`bap_<tenant>_*`)
```
curl -X DELETE <OSD URL>/api/_bap/tenant/<TENANT NAME> \
--header 'osd-xsrf: true' \
--user '<USERNAME>:<PASSWORD>' \
--header 'Content-Type: application/json' \
-d '{
  "deleteIndices": true
}'
```

sduenas commented:
> The `deleteIndices` option works perfectly
> 
> I think it would be a good option to add the indices to remove because by default the roles allow these indices schemas (the `deleteIndices` option only removes `bap_<tenant>_*`, this should be the by default behavior)
> 
>     * `grimoirelab_<tenant>_*`
> 
>     * `bap_<tenant>_*`
> 
>     * `custom_<tenant>_*`
> 
>     * `c_<tenant>_*`
> 
> 
> For example
> 
>     * With indices
> 
> 
> ```
> curl -X DELETE <OSD URL>/api/_bap/tenant/<TENANT NAME> \
> --header 'osd-xsrf: true' \
> --user '<USERNAME>:<PASSWORD>' \
> --header 'Content-Type: application/json' \
> -d '{
>   "deleteIndices": true,
>   "indices": "bap_<tenant>_*,c_<tenant>_*,git_enriched"
> }'
> ```
> 
>     * By default (`bap_<tenant>_*`)
> 
> 
> ```
> curl -X DELETE <OSD URL>/api/_bap/tenant/<TENANT NAME> \
> --header 'osd-xsrf: true' \
> --user '<USERNAME>:<PASSWORD>' \
> --header 'Content-Type: application/json' \
> -d '{
>   "deleteIndices": true
> }'
> ```

I don't like the idea of giving a list of indices with patterns to remove. Instead, it should remove all the indices related to the tenant.
